staking-pool/README.md

```diff
@@ -1,11 +1,131 @@
+The minimum epoch height when the account can withdraw is set to the current epoch height increased by `4`.
+The contract decreases the total number of staked tokens and the total number of "stake" shares. Then the contract restakes.
+
+#### Withdraw
+
+When an account wants to withdraw, the contract checks the minimum epoch height of this account and checks the amount.
+Then sends the transfer and decreases the unstaked balance of the account.
+
+#### Ping
+
+Calls the internal function to distribute rewards if the blockchain epoch switched. The contract will restake in this case.
+
+### Reward distribution
+
+Before every action the contract calls method `internal_ping`.
+This method distributes rewards towards active delegators when the blockchain epoch switches.
+The rewards might be given due to staking and also because the contract earns gas fee rebates for every function call.
+
+The method first checks that the current epoch is different from the last epoch, and if it's not changed exits the method.
+
+The reward are computed the following way. The contract keeps track of the last known total account balance.
+This balance consist of the initial contract balance, and all delegator account balances (including the owner) and all accumulated rewards.
+
+When the method is called the contract uses the current total account balance (without attached deposit) and the subtracts the last total account balance.
+The difference is the total reward that has to be distributed.
+
+The fraction of the reward is awarded to the contract owner. The fraction is configurable by the owner, but can't exceed 1.
+
+The remaining part of the reward is added to the total staked balance. This action increases the price of each "stake" share without
+changing the amount of "stake" shares owned by different accounts. Which is effectively distributing the reward based on the number of shares.
+
+The owner's reward is converted into "stake" shares at the new price and added to the owner's account.
+It's done similarly to `stake` method but without debiting the unstaked balance of owner's account.
+
+Once the rewards are distributed the contract remembers the new total balance.
+
+## Owner-only methods
+
+Contract owner can do the following:
+- Change public staking key. This action restakes with the new key.
+- Change reward fee fraction.
+- Vote on behalf of the pool. This is needed for the NEAR chain governence, and can be discussed in the following NEP: https://github.com/nearprotocol/NEPs/pull/62
+
+## Staking pool contract guarantees and invariants
+
+This staking pool implementation guarantees the required properties of the staking pool standard:
+
+- The contract can't lose or lock tokens of users.
+- If a user deposited X, the user should be able to withdraw at least X.
+- If a user successfully staked X, the user can unstake at least X.
+- The contract should lock unstaked funds for longer than 4 epochs.
```

Comment:
Shouldn't this be the opposite?

Reply:
What do you mean by opposite?

Reply:
I mean for no more than 4.  Should it ever lock for 5 or more?

Reply:
@willemneal You're correct about "should not"